packages/venia-ui/lib/components/CheckoutPage/PaymentInformation/creditCard.js

```diff
@@ -0,0 +1,276 @@
+import React, { useMemo } from 'react';
+import { bool, func, shape, string } from 'prop-types';
+import { useCreditCard } from '@magento/peregrine/lib/talons/CheckoutPage/PaymentInformation/useCreditCard';
+
+import combine from '../../../util/combineValidators';
+import {
+    hasLengthExactly,
+    isRequired,
+    validateRegionCode
+} from '../../../util/formValidators';
+import Country from '../../Country';
+import Region from '../../Region';
+import Checkbox from '../../Checkbox';
+import Field from '../../Field';
+import TextInput from '../../TextInput';
+import BrainTreeDropin from './brainTreeDropIn';
+import LoadingIndicator from '../../LoadingIndicator';
+import { mergeClasses } from '../../../classify';
+
+import creditCardPaymentOperations from './creditCard.gql';
+
+import defaultClasses from './creditCard.css';
+
+const STEP_DESCRIPTIONS = [
+    /**
+     * This is because step numbers start from 1
+     * but array indexes start from 0. Hence setting
+     * index 0 to `null`.
+     */
+    null,
+    'Submitting Billing Address',
+    'Submitting Billing Address',
+    'Submitting Payment Information',
+    'Submitting Payment Information',
+    'Submitting Payment Information',
+    'Submitting Payment Information'
+];
+
+const CreditCard = props => {
+    const {
+        classes: propClasses,
+        updateButtonClicked,
+        onPaymentSuccess: onSuccess,
+        onDropinReady: onReady,
+        brainTreeDropinContainerId,
+        onPaymentError: onError,
+        resetUpdateButtonClicked
+    } = props;
+
+    const classes = mergeClasses(defaultClasses, propClasses);
+
+    const talonProps = useCreditCard({
+        onSuccess,
+        onReady,
+        onError,
+        updateButtonClicked,
+        resetUpdateButtonClicked,
+        ...creditCardPaymentOperations
+    });
+    const {
+        shouldRequestPaymentNonce,
+        onPaymentError,
+        onPaymentSuccess,
+        onPaymentReady,
+        isBillingAddressSame,
+        countries,
+        isDropinLoading,
+        errors,
+        /**
+         * `stepNumber` depicts the state of the process flow in credit card
+         * payment flow.
+         *
+         * `0` No call made yet
+         * `1` Billing address mutation intiated
+         * `2` Billing address mutation completed
+         * `3` Braintree nonce requsted
+         * `4` Braintree nonce received
+         * `5` Payment information mutation intiated
+         * `6` Payment information mutation completed
+         * `7` All mutations done
+         */
+        stepNumber,
+        initialValues,
+        shippingAddressCountry
+    } = talonProps;
+
+    const isLoading = isDropinLoading || (stepNumber >= 1 && stepNumber <= 6);
```

Comment:
Can we simplify steps into `isSubmitting`? Seems like we only use the state here and that's all it means, right?

Reply:
It is kind of opinionated. I feel we can leave it as is. We have given enough information on the top for a future developer to understand this step.

Reply:
All I'm saying is that the steps aren't used as triggers to the effect, they're just used to indicate whether to render a loading state or not. The effects are triggered based on a bunch of state within `useCreditCard`, which is also fine, but just makes the existence of `step` kind of moot. You may as well just call `setLoading(true)` or `setLoading(false)` and  leave everything else since the comments above the effects already explain that they are happening in some order.

Reply:
I had this conversation with @jimbo a while ago. He preferred step logic. Also, it is not just to render loading. If you check, we change the text inside the loading indicator based on the step giving more info to the user about what is happening. Not totally needed but as a by-product of using steps. @jimbo had plans of using the step logic in the future. That is also the reason why the steps are more granular than needed.

Reply:
Done, there should only be 4 steps in the credit card component.

Reply:
Thanks!